dali/kernels/reduce/mean_stddev_gpu_impl.cuh

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_REDUCE_MEAN_STDDEV_GPU_IMPL_CUH_
+#define DALI_KERNELS_REDUCE_MEAN_STDDEV_GPU_IMPL_CUH_
+
+#include "dali/kernels/reduce/reduce_gpu_impl.cuh"
+#include "dali/kernels/reduce/reduce_drop_dims.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+template <typename Out, typename Scale>
+struct ScaleAndConvert {
+  Scale scale = 1;
+  template <typename T>
+  DALI_HOST_DEV Out operator()(T x) const {
+    return ConvertSat<Out>(x * scale);
+  }
+};
+
+template <typename Out, typename In, typename Actual>
+class MeanImplBase {
+ public:
+  Actual &This() { return static_cast<Actual&>(*this); }
+  const Actual &This() const { return static_cast<const Actual&>(*this); }
+
+  using scale_t = std::conditional_t<std::is_same<Out, double>::value, double, float>;
+  using Postprocessor = ScaleAndConvert<Out, scale_t>;
+
+  Postprocessor *GetPostprocessorsImpl(WorkArea &wa) const {
+    assert(!This().ReduceBatch());
+    int n = This().SimplifiedOutputShape().num_samples();
+    Postprocessor *pp = wa.ParamBuffer<Postprocessor>(n);
+    for (int i = 0; i < n; i++) {
+      DALI_ENFORCE(This().ReducedElements(i) > 0, "Cannot calculate a mean from 0 elements");
+      pp[i].scale = 1.0 / This().ReducedElements(i);
+    }
+    return pp;
+  }
+
+  Postprocessor GetPostprocessorImpl() const {
```

Comment:
When reading the comment above GetPostprocessor, there is:
```
   * Note the is_first/is_last compile-time parameters. Overload resolution will pick the
   * right function with distinct return type.
```
but actually there are no such arguments (in reduce_gpu_impl.cuh), instead there are `do_pre/postprocess`

Reply:
Also, the comment there deserves the mention, that one is used for per-sample reductions and the other for global/batch reduction.

At first I was about to comment, that they do the same thing twice in a slightly different way (and a loop). 
Still, you need to look at the implementation to know (or maybe I didn't read enough of reduce_gpu_impl).

Reply:
The comment is stale. I'll update it.

Reply:
Done.